### PR TITLE
TNT-44322 TNT-44442 Support AJO surfaces and propositions

### PR DESCRIPTION
**Pull request by XDex**
## Description

Support Surfaces in personalization query requests and AJO proposition rendering
The first AJO Beta release will only support rendering of MPA page-wide propositions (SPA support will be added in subsequent releases)
Functional e2e tests and test sandbox page added in https://jira.corp.adobe.com/browse/TNT-45510

## Related Issues

https://jira.corp.adobe.com/browse/TNT-44322
https://jira.corp.adobe.com/browse/TNT-44442
https://jira.corp.adobe.com/browse/TNT-45303
https://jira.corp.adobe.com/browse/TNT-45510

## Motivation and Context

Support Surfaces in personalization query requests and AJO proposition rendering
The first AJO Beta release will only support rendering of MPA page-wide propositions (SPA support will be added in subsequent releases)

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Improvement (non-breaking change which does not add functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] I have signed the [Adobe Open Source CLA](https://opensource.adobe.com/cla.html) or I'm an Adobe employee.
- [x] I have made any necessary test changes and all tests pass.
- [x] I have run the Sandbox successfully.

### CODE REVIEW

File: src/components/Personalization/createPersonalizationDetails.js [hunk 1]
```diff
@@ -20,8 +20,19 @@ import {
   REDIRECT_ITEM
 } from "./constants/schema";
 import isNonEmptyString from "../../utils/isNonEmptyString";
+import { WEB_SURFACE } from "./constants/surface";
+import getPageLocation from "./utils/getPageLocation";
 
-export default ({ renderDecisions, decisionScopes, event, viewCache }) => {
+const buildPageSurface = window => WEB_SURFACE + getPageLocation(window);
```
Comment (jonsnyder) [inline src/components/Personalization/createPersonalizationDetails.js:26]: Rather than passing window everywhere, you could pass this function that was already injected with window. i.e.
`components/Personalization/createBuildPageSurface.js`

```
import { WEB_SURFACE } from "./constants/surface";

export default ({ window }) => () => {
  const url = win.location.href;
  return WEB_SURFACE + url.substring(url.indexOf(":") + 3);
};
```

File: src/components/Personalization/groupDecisions.js [hunk 2]
```diff
@@ -86,14 +88,21 @@ const isViewScope = scopeDetails =>
   scopeDetails.characteristics.scopeType &&
   scopeDetails.characteristics.scopeType === VIEW_SCOPE_TYPE;
 
-const extractDecisionsByScope = decisions => {
+const isSurfaceScope = scope => scope.indexOf(WEB_SURFACE) !== -1;
+
+// only exactly matching page surfaces supported in initial AJO Beta release
+const matchesPage = (win, surface) => surface === WEB_SURFACE + getPageLocation(win);
+
+const isMatchingSurfaceScope = (win, scope) => isSurfaceScope(scope) && matchesPage(win, scope);
```
Comment (jonsnyder) [inline src/components/Personalization/groupDecisions.js:96]: The hash part of the page url could have changed in the time between request and processing. Would that be okay?

**Review by jonsnyder — COMMENTED** [review]
As we start supporting AJO and Target, I'd like to keep things better separated and more modular. This code will turn into a huge mess if we don't. We could even separate VEC and form-based offers. Certainly there is code that can be shared between AJO and Target, but lets try to make it so the differences are separated.

Here is one idea on how to make this more modular. Let me know what you think.

Build a chain of renderers or processors that handle the returned propositions. ie
- Redirect handler
- __view__ scope Form-based handler
- __view__ scope VEC handler
- SPA scope handler
- AJO surface handler

We could also have a set of functions that modify the request which could add the surface.

File: src/components/Personalization/groupDecisions.js (hunk 2 above)
Comment (XDex) [reply src/components/Personalization/groupDecisions.js:96]: > The hash part of the page url could have changed in the time between request and processing. Would that be okay?

@jonsnyder As the initial AJO Beta release will only support MPA surfaces, for SPA use cases content will be properly rendered only for the first SPA page matching the surface URI. There won't be any re-rendering on subsequent client-side SPA view changes.

**Review by ninaceban — CHANGES_REQUESTED** [review]
Thank you, Andrei!
Here are a few suggestions that I have:
- Surface should be part of the `scopeDetails` xdm schema, so that it gets collected in Platform
- Even if the returned proposition is coming from AJO, it should have `__view__` scope so that Alloy doesn't have to know about AJO. It will continue to auto render the page-wide scope, DOM_ACTION schema propositions that are coming from AJO (in future we could support SPA view scope )

File: src/components/DataCollector/index.js [hunk 3]
```diff
@@ -62,7 +63,8 @@ const createDataCollector = ({ eventManager }) => {
 
           return eventManager.sendEvent(event, {
             renderDecisions,
-            decisionScopes
+            decisionScopes,
+            surfaces
```
[the remark this answers is not in the document]
Comment (XDex) [reply src/components/DataCollector/index.js:67]: @ninaceban Yes, Erik asked for us to support this from the get-go.

File: src/components/Personalization/createGetPageLocation.js [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+
+Unless required by applicable law or agreed to in writing, software distributed under
+the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
+OF ANY KIND, either express or implied. See the License for the specific language
+governing permissions and limitations under the License.
+*/
+
+export default ({ window }) => () => {
+  const url = window.location.href;
+  return url.substring(url.indexOf(":") + 3);
```
[the remark this answers is not in the document]
Comment (XDex) [reply src/components/Personalization/createGetPageLocation.js:15]: @ninaceban This is rather an edge-case, since very few if any customers would be running their prod sites on non-standard ports, but yes, I think the port should be part of surface URI for the web channel.

File: src/components/Personalization/createPersonalizationDetails.js (hunk 1 above)
Comment (XDex) [reply src/components/Personalization/createPersonalizationDetails.js]: Fixed

**Comment by XDex** [comment]
@ninaceban Thank you for the review!
Regarding Artur's suggestion on having a separate `surface` proposition field, and have `scope` set to `__view__` for page-wide surfaces, I have detailed it [in the wiki](https://wiki.corp.adobe.com/pages/viewpage.action?pageId=2624577900), but we still need to approve this approach with Erik and Pete. Will implement the mods, once these are green-lit.

File: src/components/Personalization/createGetPageLocation.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (XDex) [reply src/components/Personalization/createGetPageLocation.js]: port is part of the surface id, as per [Surfaces pattern matching spec](https://wiki.corp.adobe.com/display/CJM/Surfaces+pattern-matching+implementation) 
Also added a test for this

**Comment by XDex** [comment]
@artur-ciocanu @ninaceban @jonsnyder I've consolidated and updated the PR according to the latest requirements documented in [AJO Web SDK integration wiki](https://wiki.corp.adobe.com/display/CJM/AJO+Web+Channel+-+Web+SDK+integration#AJOWebChannelWebSDKintegration-Conclusion). Please take another look when you have some spare time. Thanks!

**Comment by 5amfung** [comment]
@XDex Is this PR in good state that @jonsnyder can approve it in the comment here?  I know this won't be merged as we'll have a separate build for alpha testing.  But it would be great to have some status to reflect that we're complete.

**Comment by XDex** [comment]
@jonsnyder @ninaceban I have implemented the latest requested mods to sendEvent API, as per Nina's wiki and [TNT-45303](https://jira.corp.adobe.com/browse/TNT-45303), please take another look when you get a chance. Thanks!
Note: ES/EDC/AJO backend should be finally available in a staging env soon, so I'm hoping to finally be able to also add a functional test sometime this week..
cc: @5amfung @jfkhoury

**Comment by carterworks** [comment]
What is the difference between the top-level `decisionScopes` and `personalization.decisionScopes`? Is the top-level for Target and the other for AJO? Are they identical?

File: sandbox/public/perf/personalization.html [hunk 5]
```diff
@@ -34,7 +34,9 @@
       const onSendEvent = () => {
         alloy("sendEvent", {
           renderDecisions: true,
```
Comment (ninaceban) [inline sandbox/public/perf/personalization.html:36]: I would create a separate page and leave this one as is.

File: sandbox/src/PersonalizationFormBased.js [hunk 6]
```diff
@@ -13,7 +13,10 @@ const metadata = {
 
 const usePropositions = () => {
   const [propositions, setPropositions] = useState(undefined);
```
Comment (ninaceban) [inline sandbox/src/PersonalizationFormBased.js:15]: Same here, I would have a dedicated page for the new `personalization -> surface` functionality. We could deprecate them in a separate PR.

File: sandbox/src/personalizationAnalyticsClientSideHelper.js [hunk 7]
```diff
@@ -64,14 +65,17 @@ const sendEvent = ({
...
     renderDecisions,
     decisionScopes,
+    personalization,
     xdm
   });
 };
 
 export const getFormBasedOffer = () => {
   sendEvent({
     eventType: "form-based-offer",
```
Comment (ninaceban) [inline sandbox/src/personalizationAnalyticsClientSideHelper.js:75]: Same here, let's leave everything as is.

File: sandbox/src/useSendPageViewEvent.js [hunk 8]
```diff
@@ -19,7 +19,9 @@ export default ({
   viewName,
   data = {},
   xdm = {},
```
Comment (ninaceban) [inline sandbox/src/useSendPageViewEvent.js:21]: Same here.

File: src/components/DataCollector/validateUserEventOptions.js [hunk 9]
```diff
@@ -27,6 +27,10 @@ export default ({ options }) => {
     data: objectOf({}),
     renderDecisions: boolean(),
     decisionScopes: arrayOf(string()).uniqueItems(),
```
Comment (ninaceban) [inline src/components/DataCollector/validateUserEventOptions.js:29]: I am wondering if we shouldn't make `decisionScopes` and `personalization` mutually exclusive

File: src/components/Personalization/createPersonalizationDetails.js [hunk 10]
```diff
@@ -10,7 +10,8 @@ OF ANY KIND, either express or implied. See the License for the specific languag
 governing permissions and limitations under the License.
 */
 
-import { includes } from "../../utils";
+import { includes, isNonEmptyString, isNonEmptyArray } from "../../utils";
```
Comment (ninaceban) [inline src/components/Personalization/createPersonalizationDetails.js:13]: Any reason we can't have `buildPageSurface, normalizeSurfaces` directly in `./utils`? You can still have a separate file, but it is easier to control everything through a top level file/module.

File: src/components/Personalization/groupDecisions.js [hunk 11]
```diff
@@ -11,6 +11,7 @@ governing permissions and limitations under the License.
 */
 
 import { isNonEmptyArray, includes } from "../../utils";
```
Comment (ninaceban) [inline src/components/Personalization/groupDecisions.js:13]: Same here, any reason we can't have `isPageWideSurface` in `./utils`.

File: src/components/Personalization/groupDecisions.js [hunk 12]
```diff
@@ -86,14 +87,17 @@ const isViewScope = scopeDetails =>
   scopeDetails.characteristics.scopeType &&
   scopeDetails.characteristics.scopeType === VIEW_SCOPE_TYPE;
```
Comment (ninaceban) [inline src/components/Personalization/groupDecisions.js:89]: We could move this to `./utils`

File: src/components/Personalization/utils/surfaceUtils.js [hunk 13]
```diff
@@ -0,0 +1,113 @@
...
+const normalizePath = (path = "/") => {
+  let end = path.length;
+  while (end > 0 && "/".indexOf(path.charAt(end - 1)) !== -1) {
+    end -= 1;
+  }
+  return path.substring(0, end) || "/";
+};
+
+const parseSurface = surfaceString => {
+  const matched = surfaceString.match(SURFACE_REGEX);
```
Comment (ninaceban) [inline src/components/Personalization/utils/surfaceUtils.js:35]: This looks really, really, really bad. Could we use `if/else` and maybe extract some code to helper functions. Otherwise this code is pretty cryptic.

Comment (ninaceban) [reply src/components/Personalization/utils/surfaceUtils.js:35]: Also instead of calling `toLowerCase()` you could probably, make sure the initial value is lower cased.

File: src/components/Personalization/utils/surfaceUtils.js [hunk 14]
```diff
@@ -0,0 +1,113 @@
...
+    );
+  }
+  return parsed;
+};
+
+export const isPageWideSurface = scope =>
+  !!scope &&
+  scope.indexOf(WEB + SURFACE_TYPE_DELIMITER) === 0 &&
+  scope.indexOf(FRAGMENT_DELIMITER) === -1;
+
```
Comment (ninaceban) [inline src/components/Personalization/utils/surfaceUtils.js:108]: I am not following, why we need this dependency injection here. Can't we send just the `surfaces` and page location and return a complex object like {surfaces, validation errors}. The caller of this function can decide to log or do something else. Otherwise `normalizeSurfaces` looks like an utility, but does a lot of stuff including logging. Testing all this via mocking etc, should be pretty painful.

File: test/functional/specs/Personalization/C28756.js [hunk 15]
```diff
@@ -32,7 +32,9 @@ test("Test C28756: A form based offer should return if event command contains it
   const alloy = createAlloyProxy();
   await alloy.configure(config);
   const result = await alloy.sendEvent({
```
Comment (ninaceban) [inline test/functional/specs/Personalization/C28756.js:34]: Revert, let's not touch existing tests in this PR.

File: test/functional/specs/Personalization/C44363.js [hunk 16]
```diff
@@ -27,7 +27,9 @@ test("Test C44363: Return proposition when QA mode set up with token of experien
 
   const result = await alloy.sendEvent({
     renderDecisions: true,
```
Comment (ninaceban) [inline test/functional/specs/Personalization/C44363.js:29]: Same here.

File: test/functional/specs/Personalization/C44363.js [hunk 17]
```diff
@@ -46,7 +48,9 @@ test("Test C44363: Return proposition  when QA mode set up with token of experie
   });
   const result = await alloy.sendEvent({
     renderDecisions: true,
```
Comment (ninaceban) [inline test/functional/specs/Personalization/C44363.js:50]: And here.

File: test/functional/specs/Personalization/C5805676.js [hunk 18]
```diff
@@ -54,7 +54,9 @@ test("Test C5805676: Merged metric propositions should be delivered", async () =
   const alloy = createAlloyProxy();
   await alloy.configure(config);
   const eventResult = await alloy.sendEvent({
```
Comment (ninaceban) [inline test/functional/specs/Personalization/C5805676.js:56]: And here

File: test/functional/specs/Personalization/C6364799.js [hunk 19]
```diff
@@ -52,7 +52,9 @@ test("Test C6364799: applyPropositions should render html-content-item schemas w
   const alloy = createAlloyProxy();
   await alloy.configure(config);
   const result = await alloy.sendEvent({
```
Comment (ninaceban) [inline test/functional/specs/Personalization/C6364799.js:54]: And here

File: test/functional/specs/Personalization/C782718.js [hunk 20]
```diff
@@ -146,7 +146,6 @@ const simulateViewChange = async (alloy, personalizationPayload) => {
   // sendEvent at a view change, this shouldn't request any target data, it should use the existing cache
   const resultingObject = await alloy.sendEvent({
     renderDecisions: true,
```
Comment (ninaceban) [inline test/functional/specs/Personalization/C782718.js:148]: Please revert.

File: test/functional/specs/Personalization/C782719.js [hunk 21]
```diff
@@ -88,7 +88,6 @@ const simulateViewChange = async alloy => {
   // sendEvent at a view change, this shouldn't request any target data, it should use the existing cache
   const resultingObject = await alloy.sendEvent({
     renderDecisions: false,
```
Comment (ninaceban) [inline test/functional/specs/Personalization/C782719.js:90]: Please revert

File: test/unit/specs/components/DataCollector/index.spec.js [hunk 22]
```diff
@@ -60,7 +60,8 @@ describe("Event Command", () => {
       expect(event.setUserData).toHaveBeenCalledWith(data);
       expect(eventManager.sendEvent).toHaveBeenCalledWith(event, {
         renderDecisions: true,
```
Comment (ninaceban) [inline test/unit/specs/components/DataCollector/index.spec.js:62]: Reevert

File: test/unit/specs/components/DataCollector/index.spec.js [hunk 23]
```diff
@@ -69,13 +70,39 @@ describe("Event Command", () => {
   it("sends event with decisionScopes parameter when decisionScopes is not empty", () => {
     const options = {
       renderDecisions: true,
```
Comment (ninaceban) [inline test/unit/specs/components/DataCollector/index.spec.js:72]: Revert

File: test/unit/specs/components/DataCollector/index.spec.js [hunk 24]
```diff
@@ -69,13 +70,39 @@ describe("Event Command", () => {
...
-      decisionScopes: ["Foo1", "Foo2"]
+      decisionScopes: ["Foo1"],
+      personalization: {
+        decisionScopes: ["Foo2"]
+      }
     };
 
     return sendEventCommand.run(options).then(result => {
       expect(eventManager.sendEvent).toHaveBeenCalledWith(event, {
         renderDecisions: true,
```
Comment (ninaceban) [inline test/unit/specs/components/DataCollector/index.spec.js:81]: Revert

File: test/unit/specs/components/DataCollector/index.spec.js [hunk 25]
```diff
@@ -91,7 +118,8 @@ describe("Event Command", () => {
     return sendEventCommand.run({}).then(() => {
       expect(eventManager.sendEvent).toHaveBeenCalledWith(event, {
         renderDecisions: false,
```
Comment (ninaceban) [inline test/unit/specs/components/DataCollector/index.spec.js:120]: Revert

File: test/unit/specs/components/Personalization/createApplyPropositions.spec.js [hunk 26]
```diff
@@ -53,13 +53,13 @@ describe("Personalization::createApplyPropositions", () => {
     executeDecisions.and.returnValue(
       Promise.resolve(PAGE_WIDE_SCOPE_DECISIONS)
     );
```
Comment (ninaceban) [inline test/unit/specs/components/Personalization/createApplyPropositions.spec.js:55]: Please changes in this file. It just adds noise. If you want to fix it add a separate PR.

File: test/unit/specs/components/Personalization/createAutoRenderingHandler.spec.js [hunk 27]
```diff
@@ -12,10 +12,14 @@ governing permissions and limitations under the License.
 
 import {
   CART_VIEW_DECISIONS,
-  PAGE_WIDE_SCOPE_DECISIONS_WITH_DOM_ACTION_SCHEMA_ITEMS,
+  PAGE_WIDE_DECISIONS_WITH_DOM_ACTION_SCHEMA_ITEMS,
   SCOPES_FOO1_FOO2_DECISIONS
 } from "./responsesMock/eventResponses";
 import createAutoRenderingHandler from "../../../../../src/components/Personalization/createAutoRenderingHandler";
+import { isPageWideSurface } from "../../../../../src/components/Personalization/utils/surfaceUtils";
```
Comment (ninaceban) [inline test/unit/specs/components/Personalization/createAutoRenderingHandler.spec.js:19]: As mentioned ensure that everything is imported form `./utils`

File: test/unit/specs/components/Personalization/createAutoRenderingHandler.spec.js [hunk 28]
```diff
@@ -12,10 +12,14 @@ governing permissions and limitations under the License.
...
+
```
Comment (ninaceban) [inline test/unit/specs/components/Personalization/createAutoRenderingHandler.spec.js:20]: Why we have this function redeclared here, when a similar is present in prod code.

File: test/unit/specs/components/Personalization/utils/surfaceUtils.spec.js [hunk 29]
```diff
@@ -0,0 +1,256 @@
...
+    expect(isPageWideSurface("__view__")).toBeFalse();
+    expect(isPageWideSurface("name")).toBeFalse();
+    expect(isPageWideSurface("web://domain.com")).toBeTrue();
+    expect(isPageWideSurface("web://domain.com/path/page.html")).toBeTrue();
+    expect(isPageWideSurface("web://domain.com#fragment")).toBeFalse();
+    expect(isPageWideSurface("webapp://domain.com")).toBeFalse();
+    expect(isPageWideSurface("webapp://domain.com#view")).toBeFalse();
+  });
+
+  it("expands fragment surfaces", () => {
```
Comment (ninaceban) [inline test/unit/specs/components/Personalization/utils/surfaceUtils.spec.js:69]: As mentioned earlier `normalizeSurfaces` could be much simpler if we do not use dep injection.

**Review by ninaceban — CHANGES_REQUESTED** [review]
@XDex I have checked your PR. While functionally it does the right thing there are a lot of changes that are not relevant to adding `personalization -> surfaces`. Let's keep the PR focused.

There are a lot of files and hard to follow, what is relevant and what is additional refactoring. I like the idea of improving code, but let's have separate PRs focused on that.

Also let's cover with tests the functionality you have added. Since adding `personalization -> surfaces` and `personalization -> decisionScopes` is backward compatible, most of the existing test should work without any changes. You could add a separate functional test with `personalization -> surfaces & decisionScopes` just to prove that this combination also works.

File: sandbox/public/perf/personalization.html (hunk 5 above)
Comment (XDex) [reply sandbox/public/perf/personalization.html:36]: @ninaceban I've made changes to the sandbox and the tests to prioritize the new `personalization` config, as in the meeting we had on this change it was decided that going forward it is now the recommended way to set `decisionScopes`, and the top-level `decisionScopes` config is now a deprecated one. So, anyone looking through the code/test/sandbox should be aware of this.
@jfkhoury please let me know if my understanding is wrong and we should revert all the mods relating to this.

File: src/components/Personalization/createPersonalizationDetails.js (hunk 10 above)
Comment (XDex) [reply src/components/Personalization/createPersonalizationDetails.js:13]: @ninaceban Yes, in a previous comment @jonsnyder requested that we have all surface-related utils encapsulated in a separate module.

File: src/components/Personalization/utils/surfaceUtils.js (hunk 13 above)
Comment (XDex) [reply src/components/Personalization/utils/surfaceUtils.js:35]: Sure, I've extracted helper functions
Regarding `toLowerCase()` - I can't make the whole initial value lower case, since path and fragment should still remain case-sensitive

File: src/components/Personalization/utils/surfaceUtils.js (hunk 14 above)
Comment (XDex) [reply src/components/Personalization/utils/surfaceUtils.js:108]: This is so we're not leaking any surface normalization-related errors to the caller. Do we really need to handle surface related valdation error logging in `createPersonalizationDetails.createQueryDetails`? At that level, we're only interested in getting valid and normalized surfaces and shouldn't be handling any validation-specific errors.
Testing is also not a problem, as we can just mock `logger.warn` as I've done in the tests

File: test/unit/specs/components/Personalization/createAutoRenderingHandler.spec.js (hunk 28 above)
Comment (XDex) [reply test/unit/specs/components/Personalization/createAutoRenderingHandler.spec.js:20]: I figured it doesn't make sense to export internal `isPageWideScope` from `groupDecisions` if it's just used in a test and not in actual code.

File: sandbox/src/PersonalizationFormBased.js (hunk 6 above)
Comment (XDex) [reply sandbox/src/PersonalizationFormBased.js:15]: As I mentioned in the previous comment reply, we do need to also modify all existing sandbox pages to prioritize the new `personalization -> decisionScopes` format. However, as part of AJO tests, I will also be adding a dedicated sandbox page for AJO

File: test/unit/specs/components/Personalization/utils/surfaceUtils.spec.js (hunk 29 above)
Comment (XDex) [reply test/unit/specs/components/Personalization/utils/surfaceUtils.spec.js:69]: Replied in the earlier comment

File: sandbox/public/perf/personalization.html (hunk 5 above)
Comment (ninaceban) [reply sandbox/public/perf/personalization.html:36]: In the current PR we should only add the `personalization` config to support surfaces, in a following PR we will address the deprecation strategy. Even though we are recommending that the new way of querying personalized content is by sending the `decisionScopes` in the `personalization` config. We still need to be able to validate that the existing functionality works properly. That is why it is so important that we leave the sandbox app and functional tests as are, and add new page/functional tests to test the new behavior.
cc: @jfkhoury

File: src/components/Personalization/utils/surfaceUtils.js (hunk 13 above)
Comment (ninaceban) [reply src/components/Personalization/utils/surfaceUtils.js:42]: is it the fragment case-sensitive at authoring mode too?

File: src/components/Personalization/utils/surfaceUtils.js [hunk 30]
```diff
@@ -0,0 +1,113 @@
...
+          ? matched[1].toLowerCase()
+          : "",
+        authority: isNonEmptyString(matched[2]) ? matched[2].toLowerCase() : "",
+        path: isNonEmptyString(matched[3]) ? normalizePath(matched[3]) : "/",
+        fragment: matched[4]
+      }
+    : null;
+};
+
+const stringifySurface = surface =>
```
[the remark this answers is not in the document]
Comment (ninaceban) [reply src/components/Personalization/utils/surfaceUtils.js:53]: @XDex in this repo we favor string templates, even if other `stackoverflow` users might not share this view.

File: test/unit/specs/components/Personalization/createAutoRenderingHandler.spec.js (hunk 28 above)
Comment (ninaceban) [reply test/unit/specs/components/Personalization/createAutoRenderingHandler.spec.js:20]: add it in `Personalization/utils` and reuse it in both places.

File: src/components/Personalization/utils/surfaceUtils.js [hunk 31]
```diff
@@ -0,0 +1,118 @@
...
+    end -= 1;
+  }
+  return path.substring(0, end) || "/";
+};
+
+const getSurfaceType = surfaceTypeMatch =>
+  isNonEmptyString(surfaceTypeMatch) ? surfaceTypeMatch.toLowerCase() : "";
+const getAuthority = authorityMatch =>
+  isNonEmptyString(authorityMatch) ? authorityMatch.toLowerCase() : "";
+const getPath = pathMatch =>
```
Comment (ninaceban) [inline src/components/Personalization/utils/surfaceUtils.js:38]: and here as well

File: sandbox/public/perf/personalization.html (hunk 5 above)
Comment (jfkhoury) [reply sandbox/public/perf/personalization.html:36]: This comment addresses all the `Sandbox` related comments:

We need to tackled to goals here:

1. Test the new `personalization` option.
2. Confirm that the `decisionsScopes` option still works, until we decide to deprecate it. (By the way we need a ticket to do so in v 3.0)

So I suggest we do the following:

1. I agree with @ninaceban, let's leave these pages alone for now; maybe add an inline comment that this `decisionScopes` option will be deprecated in v 3.0.
2. Let's add a brand new AJO Surfaces/Web Channel Sandbox page that showcases the surfaces support and use the new `personalization` option.
3. I would like to see a test/sandbox page/something that showcases the usage of `decisionsScopes` and `personalization.decisionsScopes` and see what happens.

Thank you @XDex !!

Comment (XDex) [reply sandbox/public/perf/personalization.html:36]: @jfkhoury It's probably a misunderstanding on my end of the decisions taken in the meeting we had on nesting `decisionScopes` in `personalization` - I thought we'll be deprecating the top-level `decisionScopes` in my PR and not a separate one. Let me revert the sandbox mods and add inline comments to the API as you suggest.
Regarding p.2 - I've already added two e2e functional tests for AJO and a dedicated page to the sandbox, please take a look.
On p.3 - I've also previously added a test where we're checking that the top-level `decisionScopes` option is working as expected, either standalone or together with the new `personalization.decisionScopes`

File: sandbox/src/PersonalizationFormBased.js (hunk 6 above)
Comment (XDex) [reply sandbox/src/PersonalizationFormBased.js:15]: Reverted

File: sandbox/src/personalizationAnalyticsClientSideHelper.js (hunk 7 above)
Comment (XDex) [reply sandbox/src/personalizationAnalyticsClientSideHelper.js:75]: Reverted

File: sandbox/src/useSendPageViewEvent.js (hunk 8 above)
Comment (XDex) [reply sandbox/src/useSendPageViewEvent.js:21]: Reverted

File: src/components/DataCollector/validateUserEventOptions.js (hunk 9 above)
Comment (XDex) [reply src/components/DataCollector/validateUserEventOptions.js:29]: I don't think we should do that, since we want to maintain backwards compatibility at least initially. If we do decide to enforce mutual exclusivity, it should be done as part of a future task for deprecating the top-level `decisionScopes` option.

File: src/components/Personalization/groupDecisions.js (hunk 11 above)
Comment (XDex) [reply src/components/Personalization/groupDecisions.js:13]: Yes, see previous comment.

File: src/components/Personalization/groupDecisions.js (hunk 12 above)
Comment (XDex) [reply src/components/Personalization/groupDecisions.js:89]: Yes, but Jon previously suggested we keep all surface-related utils in a single module

File: src/components/Personalization/utils/surfaceUtils.js [hunk 32]
```diff
@@ -0,0 +1,118 @@
...
+  let end = path.length;
+  while (end > 0 && "/".indexOf(path.charAt(end - 1)) !== -1) {
+    end -= 1;
+  }
+  return path.substring(0, end) || "/";
+};
+
+const getSurfaceType = surfaceTypeMatch =>
+  isNonEmptyString(surfaceTypeMatch) ? surfaceTypeMatch.toLowerCase() : "";
+const getAuthority = authorityMatch =>
```
[the remark this answers is not in the document]
Comment (XDex) [reply src/components/Personalization/utils/surfaceUtils.js:36]: Fixed

File: src/components/Personalization/utils/surfaceUtils.js (hunk 31 above)
Comment (XDex) [reply src/components/Personalization/utils/surfaceUtils.js:38]: Fixed

File: src/components/Personalization/utils/surfaceUtils.js (hunk 13 above)
Comment (XDex) [reply src/components/Personalization/utils/surfaceUtils.js:42]: Fragments are not yet supported in the UI, but they are expected to be case-sensitive, yes

File: src/components/Personalization/utils/surfaceUtils.js (hunk 30 above)
[the remark this answers is not in the document]
Comment (XDex) [reply src/components/Personalization/utils/surfaceUtils.js:53]: @ninaceban Hmm, I see that in `utils/injectStorage` for example, string concatenation is used.. But let me replace it with string templates as you suggest.

File: test/functional/specs/Personalization/C28756.js (hunk 15 above)
Comment (XDex) [reply test/functional/specs/Personalization/C28756.js:34]: Reverted

File: test/functional/specs/Personalization/C44363.js (hunk 16 above)
Comment (XDex) [reply test/functional/specs/Personalization/C44363.js:29]: Reverted

File: test/functional/specs/Personalization/C44363.js (hunk 17 above)
Comment (XDex) [reply test/functional/specs/Personalization/C44363.js:50]: Done

File: test/functional/specs/Personalization/C5805676.js (hunk 18 above)
Comment (XDex) [reply test/functional/specs/Personalization/C5805676.js:56]: Done

File: test/functional/specs/Personalization/C6364799.js (hunk 19 above)
Comment (XDex) [reply test/functional/specs/Personalization/C6364799.js:54]: Done

File: test/functional/specs/Personalization/C782718.js (hunk 20 above)
Comment (XDex) [reply test/functional/specs/Personalization/C782718.js:148]: Done

File: test/functional/specs/Personalization/C782719.js (hunk 21 above)
Comment (XDex) [reply test/functional/specs/Personalization/C782719.js:90]: Done

File: test/unit/specs/components/DataCollector/index.spec.js (hunk 22 above)
Comment (XDex) [reply test/unit/specs/components/DataCollector/index.spec.js:62]: I have reverted the changes related to the new option from the API and the sandbox, but as @jfkhoury mentioned in a previous comment, we need these to be present in tests, in order to confirm that both the top-level `decisionScopes` and the new `personalization.decisionScopes` options are working as expected.

File: test/unit/specs/components/Personalization/createApplyPropositions.spec.js (hunk 26 above)
Comment (XDex) [reply test/unit/specs/components/Personalization/createApplyPropositions.spec.js:55]: @ninaceban It's just a minor fix, can we please leave it?

File: test/unit/specs/components/Personalization/createAutoRenderingHandler.spec.js (hunk 27 above)
Comment (XDex) [reply test/unit/specs/components/Personalization/createAutoRenderingHandler.spec.js:19]: See previous comment on this

File: test/unit/specs/components/Personalization/createAutoRenderingHandler.spec.js (hunk 28 above)
Comment (XDex) [reply test/unit/specs/components/Personalization/createAutoRenderingHandler.spec.js:20]: Fixed

**Comment by XDex** [comment]
@ninaceban @jonsnyder @jfkhoury I've addressed the review comments and rebased on latest `main`. Please take another look when you get a chance. Thank you!

File: src/components/DataCollector/index.js [hunk 33]
```diff
@@ -30,7 +30,7 @@ const createDataCollector = ({ eventManager }) => {
             mergeId,
             renderDecisions = false,
             decisionScopes = [],
```
[the remark this answers is not in the document]
Comment (XDex) [reply src/components/DataCollector/index.js]: Done

**Comment by jonsnyder** [comment]
Closing in favor of #920
